utils/configutil/config.go

```diff
@@ -117,111 +103,70 @@ func (e ValidationError) Error() string {
-	var paths []string
-	cSet := make(stringset.Set)
-
-	// Go through all the 'extends' hierarchy until there is no base anymore or
-	// some reference cycles have been detected.
-	cSet.Add(fname)
-
-	candidate := getCandidate(fname, dirs)
-	fmt.Fprintf(os.Stderr, "candidate: %s\n", candidate)
-	if candidate == "" {
-		return nil, ErrNoFilesToLoad
-	}
+type getExtend func(filename string) (extends string, err error)
 
+// resolveExtends returns the list of config paths that the original config `filename`
+// points to.
+func resolveExtends(filename string, extendReader getExtend) ([]string, error) {
+	filenames := []string{filename}
+	fileSet := make(stringset.Set)
 	for {
-		extends, err := readExtend(candidate)
+		extends, err := extendReader(filename)
 		if err != nil {
 			return nil, err
+		} else if extends == "" {
+			break
 		}
 
-		paths = append([]string{candidate}, paths...)
-		if extends != "" {
-			// Prevent circular references.
-			if !cSet.Has(extends) {
-				candidate = path.Join(filepath.Dir(candidate), extends)
-				cSet.Add(extends)
-			} else {
-				return nil, ErrCycleRef
-			}
-		} else {
-			break
+		// If the file path of the extends field in the config is not absolute
+		// we assume that it is in the same directory as the current config
+		// file.
+		if !filepath.IsAbs(extends) {
+			extends = path.Join(filepath.Dir(filename), extends)
 		}
-	}
 
-	// Append secrets.
-	candidate = getCandidate(secretsFile, dirs)
-	if candidate != "" {
-		paths = append(paths, candidate)
-	}
+		// Prevent circular references.
+		if fileSet.Has(extends) {
+			return nil, ErrCycleRef
+		}
 
-	return paths, nil
+		filenames = append([]string{extends}, filenames...)
+		fileSet.Add(extends)
+		filename = extends
+	}
+	return filenames, nil
 }
 
-// Load loads configuration based on config file name.
-// If config directory cannot be derived from file name, get it from environment
-// variables.
-func Load(fname string, config interface{}) error {
-	candidates, err := filterCandidatesFromDirs(
-		filepath.Base(fname), []string{filepath.Dir(fname)})
-	if err != nil && err != ErrNoFilesToLoad {
-		return fmt.Errorf("find config under %s: %s", filepath.Dir(fname), err)
-	} else if err == ErrNoFilesToLoad {
-		candidates, err = FilterCandidates(fname)
-		if err != nil {
-			return fmt.Errorf(
-				"find config under %s and %s: %s", filepath.Dir(fname), configDirKey, err)
-		}
+func readExtend(configFile string) (string, error) {
+	data, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		return "", err
 	}
 
-	return loadFiles(config, candidates...)
+	var cfg Extends
+	if err := yaml.Unmarshal(data, &cfg); err != nil {
+		return "", fmt.Errorf("unmarshal %s: %s", configFile, err)
+	}
+	return cfg.Extends, nil
 }
 
-// LoadFiles loads a list of files, deep-merging values.
+// loadFiles loads a list of files, deep-merging values.
 func loadFiles(config interface{}, fnames ...string) error {
 	if len(fnames) == 0 {
 		return ErrNoFilesToLoad
```

Comment:
This is dead code I believe -- now we'll always have at least one fname.

Reply:
(Also maybe get rid of the variadic arguments which don't really make sense in the context of this file)

Reply:
You're right; it's dead code. Also replacing var args with slice.